### PR TITLE
[9.0][ADD] stock_removal_location_by_priority

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Stock Removal Location by Priority
==================================

This module adds a removal priority field on stock locations.
This priority applies when removing a product from different stock locations
and the incoming dates are equal in both locations.

Solves https://github.com/OCA/stock-logistics-warehouse/issues/286.

### CODE REVIEW

File: stock_removal_location_by_priority/README.rst [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+==================================
+
+This module adds a removal priority field on stock locations.
+This priority applies when removing a product from different stock locations
+and the incoming dates are equal in both locations.
+
+Configuration
+=============
+
+You can configure the rules to compute the cycle count, acting as follow:
```
Comment (LoisRForgeFlow) [inline stock_removal_location_by_priority/README.rst]: I think this line does not belong here :wink:

File: stock_removal_location_by_priority/models/stock_quant.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        if any(move.mapped('location_id.removal_priority')):
+            if removal_strategy == 'fifo':
+                order = 'in_date, removal_priority, id'
+                return self._quants_get_order(
+                    quantity, move, ops=ops, domain=domain, orderby=order)
+            elif removal_strategy == 'lifo':
+                order = 'in_date desc, removal_priority desc, id desc'
+                return self._quants_get_order(
+                    quantity, move, ops=ops, domain=domain, orderby=order)
+            raise UserError(_('Removal strategy %s not implemented.') % (
```
Comment (LoisRForgeFlow) [inline stock_removal_location_by_priority/models/stock_quant.py]: Are you sure that to raise an error is the good call here? Maybe just logging a warning message in the server logs could do the job and do not block new possible removal strategies.

File: stock_removal_location_by_priority/models/stock_quant.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    @api.model
+    def apply_removal_strategy(self, quantity, move, ops=False,
+                               domain=None, removal_strategy='fifo'):
+        if any(move.mapped('location_id.removal_priority')):
+            if removal_strategy == 'fifo':
+                order = 'in_date, removal_priority, id'
+                return self._quants_get_order(
+                    quantity, move, ops=ops, domain=domain, orderby=order)
+            elif removal_strategy == 'lifo':
+                order = 'in_date desc, removal_priority desc, id desc'
```
Comment (LoisRForgeFlow) [inline stock_removal_location_by_priority/models/stock_quant.py]: I think that removal_priority should be sorted crescent here too. No matter what removal strategy you use when the removal_priority is the one that would decide we want it to always pick the same way.

**Review by LoisRForgeFlow — COMMENTED** [review]
Some comments in line

File: stock_removal_location_by_priority/models/stock_quant.py (hunk 2 above)
Comment (MiquelRForgeFlow) [reply stock_removal_location_by_priority/models/stock_quant.py]: The base apply_removal_strategy method has this UserError. Perhaps, other possible strategies that may exist they also are included via Super like in this case.

Comment (LoisRForgeFlow) [reply stock_removal_location_by_priority/models/stock_quant.py]: Yes, you are right, better to go along with standard

**Review by LoisRForgeFlow — APPROVED** [review]

**Review by lmignon — APPROVED** [review]
@mreficent Thank you for this new module: LGTM (functional + code review)

**Review by Cedric-Pigeon — APPROVED** [review]

**Pull request merged**